spec/registry.md

```diff
@@ -0,0 +1,156 @@
+# WIP DRAFT MessageFormat 2.0 Registry
+
+_This document is non-normative._
+
+The implementations and tooling can greatly benefit from a structured definition of formatting and matching functions available to messages at runtime.
+The _registry_ is a mechanism for storing such declarations in a portable manner.
+
+## Goals
+
+The goal of the registry is to provide a machine-readable description of MessageFormat extensions (custom functions) in order to:
+
+* support the localization roundtrip,
+* validate message correctness, and
+* improve the authoring experience.
+
+## Use Cases
+
+The registry enables the following usage scenarios:
+
+* Generate variant keys for a given locale during XLIFF extraction.
+* Verify the exhaustiveness of variant keys given a selector.
+* Type-check values passed into functions.
+* Validate that matching functions are only called in selectors.
+* Validate that formatting functions are only called in placeholders.
+* Forbid edits to certain function options (e.g. currency options).
+* Autocomplete function and option names.
+* Display on-hover tooltips for function signatures with documentation.
+* Display/edit known message metadata.
+* Restrict input in GUI by providing a dropdown with all viable option values.
+
+## Data Model
+
+The registry contains descriptions of function signatures.
+[`registry.dtd`](./registry.dtd) describes its data model.
+
+The main building block of the registry is the `<function>` element.
+It represents an implementation of a custom function available to translation at runtime.
+A function defines a human-readable _description_ of its behavior
+and one or more machine-readable _signatures_ of how to call it.
+Named `<pattern>` elements can optionally define regex validation rules for input, option values, and variant keys.
+
+A _signature_ defines one particular set of arguments and named options that can be used together in a single call to the function.
+`<formatSignature>` corresponds to a function call inside a placeholder inside translatable text.
+`<matchSignature>` coresponds to a function call inside a selector.
+Signatures with a non-empty `locales` attribute are locale-specific and only available in translations in the given languages.
+
+A signature may define the positional argument of the function with the `<input>` element.
+A signature may also define one or more `<option>` elements representing _named options_ to the function.
+Options are optional by default,
+unless the `required` attribute is present.
+They accept either a finite enumeration of values (the `values` attribute)
+or validate they input with a regular expression (the `pattern` attribute).
+Read-only options (the `readonly` attribute) can be displayed to translators in CAT tools, but may not be edited.
+
+Matching-function signatures additionally include one or more `<match>` elements
```

Comment:
Is it "one or more" or "zero or more"? The DTD implies that there can be zero `<match>` elements, as does the second signature for `number` in the example below.